### DISCUSSION TITLE
does a matrix build share somehow resources, caches, deamons?

### DISCUSSION THREAD

**Discussion by robstoll** [community/community#49223] [Actions]
### Select Topic Area

Bug

### Body

I periodically (unfortunately not rarely) run into an issue in a matrix build (using jdk 11 and 14) where the kotlin compilation fails unexpectedly.
The error never happens locally and does never happen if I reduce the build to just one jdk version. 
Hence, my suspicion, that the builds somehow share something. Does someone know if there are such optimisations in place or should the builds in theory be completely separated?
I already run into a similar issue some years ago where one build's package-lock.json file caused interference with another build's package-lock.json where yarn was used. One reason more, that I suspect github somehow has an optimisation in place which might work most of the time but does not for me :cry: 

It's quite cumbersome to workaround this bug (feature), so I hope to get at least a confirmation that this is intentional.

Following an example where it happened:
https://github.com/robstoll/tutteli-gradle-plugins/actions/runs/4317517440/jobs/7534676181#step:4:206
The other build involved in this matrix build: https://github.com/robstoll/tutteli-gradle-plugins/actions/runs/4317517440/jobs/7534676072#step:4:186

**Comment by jsoref** [comment]
My guess is that the [actions/setup-java](https://github.com/actions/setup-java) code is not being careful to avoid mixing java versions... afaict, the code it's using is more or less near here:
https://github.com/actions/setup-java/blob/ea15b3b99cdc9ac45af1882d085e3f9297a75a8b/src/cleanup-java.ts#L23

You have choices:
1. you could change the workflow not not use the action
2. you could change the workflow to ask the action not to use caches
3. you could file a bug to the action (pointing to the same things you did here)
4. you could file a bug to support.github.com (approximately equivalent to 3)
5. you could fix the action's poor cache handling

To answer the question you technically asked, "no, matrix doesn't naturally share more things than not using a matrix, but if you or something you use tries hard enough it will let you do stupid things". Slightly more precisely, "the set of caches available for a given commit for a given branch are accessible to all workflows run from that same branch and commit in that same repository from the same PR / push". (Two repositories with the same branch name pointing to the same commit would not share caches at all, in general, two PRs from distinct repositories to a common repository using the same source and destination branches should also not share interesting caches, but might share common read-only ancestor caches.)

Matrix aren't going to share daemons any more than non matrix builds. Workflows of course do share general github services (upload-artifact, download-artifact, cache, git repositories).

Is it intentional? I mean, someone wrote the code they wrote. My bet is they didn't consider enough corner cases. Whether you should continue to use that component would be up to you.

https://github.com/actions/setup-java/issues/459
https://github.com/actions/setup-java/issues/389 -- this argues that you shouldn't use the action at all.
https://github.com/actions/setup-java/issues/84 -- this should have been resolved when they clearly added the code to use actions/cache but it doesn't look like they're doing a good job of maintaining their repository.

A quick skim through https://github.com/actions/setup-java/issues shows that the repository is relatively unhealthy. They aren't doing a good job of resolving tickets once they implement the requested items (and it isn't like they have millions of tickets -- I used to wrangle much larger bug databases).

**Reply by robstoll** [reply]
Thanks for the answer, it's particularly good to know that

> the set of caches available for a given commit for a given branch are accessible to all workflows run from that same branch and commit in that same repository from the same PR / push

I guess the location where the cache is written can already be shared before the upload happens? 
- If not, then it cannot be the problem because neither of the two jobs have been successful yet, i.e. the cache should not have been uploaded yet. 
- if it is shared, then I guess we found the problem. gradle (or the kotlin compiler, not sure which one is the culprit) reads a compiled module-info.class which was compiled by the other build which happen to have just compiled the failure case and the other one just wants to run the test for the happy case (or vice versa).

**Reply by jsoref** [reply]
There are two ways to interact with actions/cache. One is by using the actions/cache action which will only write to the cache upon workflow success, and the other is to do black magic. actions/setup-java is doing the latter which means that it can definitely write before completion.
